dbms/src/Flash/Mpp/MPPTunnel.cpp

```diff
@@ -195,32 +189,63 @@ void MPPTunnel::writeDone()
 void MPPTunnel::connect(PacketWriter * writer)
 {
     {
-        std::unique_lock lk(*mu);
+        std::unique_lock lk(mu);
         if (status != TunnelStatus::Unconnected)
             throw Exception(fmt::format("MPPTunnel has connected or finished: {}", statusToString()));
 
         LOG_FMT_TRACE(log, "ready to connect");
         switch (mode)
         {
         case TunnelSenderMode::LOCAL:
+        {
             RUNTIME_ASSERT(writer == nullptr, log);
-            local_tunnel_sender = std::make_shared<LocalTunnelSender>(mode, send_queue, nullptr, log, tunnel_id);
+            local_tunnel_sender = std::make_shared<LocalTunnelSender>(queue_size, log, tunnel_id);
             tunnel_sender = local_tunnel_sender;
             break;
+        }
         case TunnelSenderMode::SYNC_GRPC:
+        {
             RUNTIME_ASSERT(writer != nullptr, log, "Sync writer shouldn't be null");
-            sync_tunnel_sender = std::make_shared<SyncTunnelSender>(mode, send_queue, writer, log, tunnel_id);
-            sync_tunnel_sender->startSendThread();
+            sync_tunnel_sender = std::make_shared<SyncTunnelSender>(queue_size, log, tunnel_id);
+            sync_tunnel_sender->startSendThread(writer);
             tunnel_sender = sync_tunnel_sender;
             break;
-        case TunnelSenderMode::ASYNC_GRPC:
-            RUNTIME_ASSERT(writer != nullptr, log, "Async writer shouldn't be null");
-            async_tunnel_sender = std::make_shared<AsyncTunnelSender>(mode, send_queue, writer, log, tunnel_id, mu);
-            tunnel_sender = async_tunnel_sender;
-            writer->attachAsyncTunnelSender(async_tunnel_sender);
-            break;
+        }
         default:
-            RUNTIME_ASSERT(false, log, "Unsupported TunnelSenderMode: {}", mode);
+            RUNTIME_ASSERT(false, log, "Unsupported TunnelSenderMode in connect: {}", mode);
+        }
+        status = TunnelStatus::Connected;
+        cv_for_status_changed.notify_all();
+    }
+    LOG_DEBUG(log, "connected");
+}
+
+void MPPTunnel::connectAsync(IAsyncCallData * call_data)
+{
+    {
+        std::unique_lock lk(mu);
+        if (status != TunnelStatus::Unconnected)
+            throw Exception(fmt::format("MPPTunnel has connected or finished: {}", statusToString()));
```

Comment:
Almost all codes of `MPPTunnel` are written for async grpc only. 
So we should totally separate the sync grpc code path from `MPPTunnel` to clarify the code.
Since this PR is too large, it will be done in future PR.